File: 0056-merge-intervals/0056-merge-intervals.ts
function merge(intervals: number[][]): number[][] {
    intervals.sort((a, b) => a[0] - b[0]);
    let merged = [intervals[0]];

    for (let i = 1; i < intervals.length; i++) {
        const n = merged.length;
        const curr = intervals[i];

        if (merged[n - 1][1] >= curr[0]) {
            merged[n - 1][1] = Math.max(curr[1], merged[n - 1][1]);
        } else {
            merged.push(curr)
        }
    }


    return merged;
};

File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        int n = intervals.length;
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        List<int[]> merged = new ArrayList<>();
        merged.add(intervals[0]);

        for (int i = 1; i < n; i++) {
            int[] prev = merged.get(merged.size() - 1);
            int[] curr = intervals[i];

            if (curr[0] <= prev[1]) {
                prev[1] = Math.max(prev[1], curr[1]);
            } else {
                merged.add(curr);
            }
        }

        return merged.toArray(new int[merged.size()][]);
    }
}